File: regelsuche-release/src/main/java/de/regelsuche/release/ReleaseReadinessBindings.java
package de.regelsuche.release;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.DeserializationFeature;
import de.regelsuche.experiments.autopilot.AutonomousProductionCampaignRunner;
import de.regelsuche.experiments.autopilot.AutonomousProductionCampaignRunner.ArtifactReference;
import de.regelsuche.release.ReleaseReadinessMatrix.MatrixReport;
import de.regelsuche.release.ReleaseReadinessMatrix.ProfileResult;
import de.regelsuche.release.ReleaseReadinessMatrix.ProfileStatus;
import de.regelsuche.release.ReleaseReadinessMatrix.RequirementCheck;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.util.ArrayList;
import java.util.EnumMap;
import java.util.HexFormat;
import java.util.List;
import java.util.Map;

/** Verifies direct retained release bindings; it never executes a campaign. */
final class ReleaseReadinessBindings {
    private ReleaseReadinessBindings() { }
    static void verify(Map<String, JsonNode> documents, byte[] catalogBytes) {
        JsonNode run = document(documents, "release-readiness-run.json");
        JsonNode matrix = document(documents, "release-readiness-report.json");
        JsonNode campaign = document(documents, "campaign/production-campaign-manifest.json");
        JsonNode evidence = document(documents, "evidence-summary.json");
        JsonNode hidden = document(documents, "hidden-rule-release-evidence.json");
        JsonNode qualification = document(documents, "qualification/candidate-qualification-evidence.json");
        JsonNode qualificationRun = document(documents, "qualification/candidate-qualification-run.json");
        JsonNode catalog = document(documents, "profiles.json");

        // The original domain constructors own normalization, accounting and canonical identity.
        decode(evidence, AutonomousCampaignReleaseEvidence.class);
        decode(hidden, HiddenRuleBenchmarkReleaseEvidence.class);
        decode(qualification, AutonomousCandidateQualificationEvidence.class);
        require(text(campaign, "schema").equals(AutonomousProductionCampaignRunner.SCHEMA), "unsupported campaign schema");
        var artifacts = new ArrayList<ArtifactReference>();
        for (JsonNode artifact : array(campaign, "artifacts")) {
            artifacts.add(new ArtifactReference(text(artifact, "artifactType"), text(artifact, "contentHash")));
        }
        require(text(campaign, "contentHash").equals(AutonomousProductionCampaignRunner.manifestHash(
            text(campaign, "briefHash"), text(campaign, "lifecycleRunHash"), text(campaign, "nextPlanHash"),
            text(campaign, "campaignRoundHash"), text(campaign, "feedbackReallocationHash"),
            text(campaign, "campaignResourceLedgerHash"), artifacts, text(campaign, "status"))), "campaign canonical identity differs");
        require(text(qualificationRun, "schema").equals(AutonomousCandidateQualificationRunner.SCHEMA), "unsupported qualification run schema");
        require(text(qualificationRun, "contentHash").equals(AutonomousCandidateQualificationRunner.runHash(
            text(qualificationRun, "campaignManifestHash"), text(qualificationRun, "suiteHash"),
            text(qualificationRun, "splitAuditHash"), text(qualificationRun, "evaluationHash"),
            text(qualificationRun, "utilityHash"), text(qualificationRun, "qualificationEvidenceHash"))), "qualification run canonical identity differs");
        for (String field : List.of("seedFamilyCount", "observationCount", "candidateCount", "rejectedClusterCount")) {
            same(campaign, field, evidence, field, "campaign summary count differs: " + field);
        }
        require(bool(evidence, "targetFree") != bool(campaign, "targetProvided"), "campaign target-free summary differs");
        for (String field : List.of("suiteHash", "splitAuditHash", "evaluationHash", "utilityHash", "qualified")) {
            same(qualificationRun, field, qualification, field, "qualification run summary differs: " + field);
        }

        require(text(run, "schema").equals(ReleaseReadinessRunner.SCHEMA), "unsupported release run schema");
        Map<String, String> expected = Map.of("campaignManifestHash", text(campaign, "contentHash"),
            "profileCatalogHash", sha256(catalogBytes), "evidenceHash", text(evidence, "evidenceHash"),
            "hiddenRuleEvidenceHash", text(hidden, "evidenceHash"), "qualificationEvidenceHash", text(qualification, "contentHash"),
            "matrixHash", text(matrix, "contentHash"));
        expected.forEach((field, hash) -> require(text(run, field).equals(hash), "run " + field + " artifact binding differs"));
        for (JsonNode dependent : List.of(evidence, qualification, qualificationRun)) {
            same(dependent, "campaignManifestHash", run, "campaignManifestHash", "dependent evidence belongs to another campaign");
        }
        same(qualificationRun, "qualificationEvidenceHash", run, "qualificationEvidenceHash", "qualification run evidence binding differs");
        same(matrix, "evidenceHash", run, "evidenceHash", "matrix campaign evidence binding differs");
        same(matrix, "hiddenRuleEvidenceHash", run, "hiddenRuleEvidenceHash", "matrix hidden-rule evidence binding differs");
        same(run, "autonomousCampaignStatus", matrix, "autonomousCampaignStatus", "run autonomy summary differs");
        same(run, "autonomyClaimAuthorized", matrix, "autonomyClaimAuthorized", "run autonomy authority differs");
        verifyMatrix(matrix, catalog);
        require(text(run, "contentHash").equals(ReleaseReadinessRunner.runHash(text(run, "profileCatalogHash"),
            text(run, "evidenceHash"), text(run, "hiddenRuleEvidenceHash"), text(run, "qualificationEvidenceHash"),
            text(run, "matrixHash"), text(run, "campaignManifestHash"))), "run canonical content hash differs");
    }

    private static void verifyMatrix(JsonNode matrix, JsonNode catalog) {
        var profiles = profiles(matrix);
        var declared = profiles(catalog);
        require(text(catalog, "schema").equals(ReleaseEvidenceProfile.CATALOG_SCHEMA), "unsupported profile catalog schema");
        require(text(catalog, "autonomyClaimProfile").equals("AUTONOMOUS_CAMPAIGN")
            && text(catalog, "externalNoveltyProfile").equals("EXTERNAL_NOVELTY_REVIEW"), "catalog claim profile binding differs");
        var results = new ArrayList<ProfileResult>();
        profiles.forEach((name, profile) -> {
            JsonNode declaration = declared.get(name);
            require(bool(declaration, "authorizesAutonomyClaim") == name.authorizesAutonomyClaim(), "catalog profile authority differs");
            same(profile, "claim", declaration, "claim", "matrix profile claim differs from catalog");
            var checks = new ArrayList<RequirementCheck>();
            for (JsonNode check : array(profile, "checks")) checks.add(decode(check, RequirementCheck.class));
            var blockers = new ArrayList<String>();
            for (JsonNode blocker : array(profile, "blockers")) {
                require(blocker.isTextual(), "profile blocker must be text");
                blockers.add(blocker.textValue());
            }
            results.add(new ProfileResult(name, ProfileStatus.valueOf(text(profile, "status")),
                bool(profile, "authorizesAutonomyClaim"), checks, blockers));
        });
        new MatrixReport(text(matrix, "schema"), text(matrix, "evidenceHash"), text(matrix, "hiddenRuleEvidenceHash"),
            results, ProfileStatus.valueOf(text(matrix, "autonomousCampaignStatus")), bool(matrix, "autonomyClaimAuthorized"),
            text(matrix, "promotionStatus"), text(matrix, "publicEvidenceStatus"), text(matrix, "contentHash"));
    }

    private static Map<ReleaseEvidenceProfile, JsonNode> profiles(JsonNode document) {
        var result = new EnumMap<ReleaseEvidenceProfile, JsonNode>(ReleaseEvidenceProfile.class);
        for (JsonNode profile : array(document, "profiles")) {
            require(result.put(ReleaseEvidenceProfile.valueOf(text(profile, "profile")), profile) == null, "duplicate release profile");
        }
        require(result.size() == ReleaseEvidenceProfile.values().length, "missing release profile");
        return result;
    }

    private static <T> T decode(JsonNode node, Class<T> type) {
        try {
            return new ObjectMapper().disable(DeserializationFeature.ACCEPT_FLOAT_AS_INT)
                .enable(DeserializationFeature.FAIL_ON_NULL_FOR_PRIMITIVES).treeToValue(node, type);
        } catch (com.fasterxml.jackson.core.JsonProcessingException error) {
            throw new IllegalArgumentException("retained " + type.getSimpleName() + " does not satisfy its original contract", error);
        }
    }

    static JsonNode document(Map<String, JsonNode> documents, String name) {
        JsonNode result = documents.get(name);
        require(result != null && result.isObject(), "missing object document: " + name);
        return result;
    }

    static JsonNode field(JsonNode object, String field) {
        JsonNode result = object.get(field);
        require(result != null, "missing field: " + field);
        return result;
    }
    static String text(JsonNode object, String name) {
        JsonNode value = field(object, name);
        require(value.isTextual(), "text field required: " + name);
        return value.textValue();
    }
    static boolean bool(JsonNode object, String name) {
        JsonNode value = field(object, name);
        require(value.isBoolean(), "boolean field required: " + name);
        return value.booleanValue();
    }
    static JsonNode array(JsonNode object, String name) {
        JsonNode value = field(object, name);
        require(value.isArray(), "array field required: " + name);
        return value;
    }
    static void same(JsonNode first, String firstField, JsonNode second, String secondField, String message) {
        require(field(first, firstField).equals(field(second, secondField)), message);
    }
    static void require(boolean valid, String message) {
        if (!valid) throw new IllegalArgumentException(message);
    }
    static String sha256(byte[] bytes) {
        try { return "sha256:" + HexFormat.of().formatHex(MessageDigest.getInstance("SHA-256").digest(bytes)); }
        catch (NoSuchAlgorithmException error) { throw new IllegalStateException("SHA-256 unavailable", error); }
    }
}

File: scripts/release_readiness_bindings.py
"""Existing ReleaseRun/MatrixReport v1 hash and artifact bindings, without replay."""

from __future__ import annotations

from pathlib import Path
from release_readiness_identities import (
    java_hash, java_list, java_string_key, sha256, verify_artifact_identities,
)

PROFILES = {
    "SEARCH_REPRODUCIBILITY", "HIDDEN_RULE_REDISCOVERY", "OPEN_TARGET_DISCOVERY",
    "AUTONOMOUS_CAMPAIGN", "EXTERNAL_NOVELTY_REVIEW",
}


def profile_material(profile: dict) -> str:
    checks = sorted(profile["checks"], key=lambda item: java_string_key(item["code"]))
    check_material = (
        item["code"] + "|passed=" + str(item["passed"]).lower()
        + "|actual=" + item["actual"] + "|required=" + item["required"]
        for item in checks
    )
    blockers = sorted(set(profile["blockers"]), key=java_string_key)
    return (
        profile["profile"] + "|" + profile["status"] + "|"
        + str(profile["authorizesAutonomyClaim"]).lower() + "|"
        + java_list(check_material) + "|" + java_list(blockers)
    )


def matrix_hash(matrix: dict) -> str:
    # ReleaseReadinessMatrix.matrixHash / ProfileResult / RequirementCheck.
    profiles = sorted(matrix["profiles"], key=lambda item: java_string_key(item["profile"]))
    material = (
        matrix["schema"] + "\nevidence=" + matrix["evidenceHash"]
        + "\nhiddenRuleEvidence=" + matrix["hiddenRuleEvidenceHash"]
        + "\nprofiles=" + java_list(profile_material(item) for item in profiles)
        + "\nautonomy=" + matrix["autonomousCampaignStatus"]
    )
    return java_hash(material)


def run_hash(run: dict) -> str:
    # ReleaseReadinessRunner.runHash: the order and labels are part of v1.
    material = (
        run["schema"] + "\nprofileCatalog=" + run["profileCatalogHash"]
        + "\nevidence=" + run["evidenceHash"]
        + "\nhiddenRuleEvidence=" + run["hiddenRuleEvidenceHash"]
        + "\nqualificationEvidence=" + run["qualificationEvidenceHash"]
        + "\nmatrix=" + run["matrixHash"]
        + "\ncampaign=" + run["campaignManifestHash"]
    )
    return java_hash(material)


def profiles_by_name(document: dict, label: str, require) -> dict:
    profiles = document.get("profiles")
    require(isinstance(profiles, list), label + " profiles must be an array")
    require(
        all(isinstance(item, dict) and isinstance(item.get("profile"), str) for item in profiles),
        label + " profiles must name their profile",
    )
    names = [item["profile"] for item in profiles]
    require(len(names) == len(PROFILES) and set(names) == PROFILES,
            label + " must contain every release profile exactly once")
    return {item["profile"]: item for item in profiles}


def verify_matrix_profiles(matrix: dict, catalog: dict, require) -> None:
    profiles = profiles_by_name(matrix, "matrix", require)
    declared = profiles_by_name(catalog, "catalog", require)
    require(catalog.get("schema") == "regelsuche.release-evidence-profile-catalog/v1",
            "unsupported profile catalog schema")
    require(catalog.get("autonomyClaimProfile") == "AUTONOMOUS_CAMPAIGN"
            and catalog.get("externalNoveltyProfile") == "EXTERNAL_NOVELTY_REVIEW",
            "catalog claim profile binding differs")
    for name, profile in profiles.items():
        authority = name == "AUTONOMOUS_CAMPAIGN"
        require(declared[name].get("authorizesAutonomyClaim") is authority,
                "catalog profile authority differs: " + name)
        require(profile["claim"] == declared[name].get("claim"),
                "matrix profile claim differs from catalog: " + name)
        ready = profile["status"] == "READY"
        require(ready == (not profile["blockers"]),
                "profile status and blockers disagree: " + name)
        require(profile["authorizesAutonomyClaim"] is (authority and ready),
                "profile claim authority is inconsistent: " + name)
    autonomy = profiles["AUTONOMOUS_CAMPAIGN"]
    require(matrix["autonomousCampaignStatus"] == autonomy["status"]
            and matrix["autonomyClaimAuthorized"] == autonomy["authorizesAutonomyClaim"],
            "matrix autonomy summary is inconsistent")


def verify_root_bindings(root: Path, load, require, read_bytes) -> None:
    run = load(root / "release-readiness-run.json")
    matrix = load(root / "release-readiness-report.json")
    campaign = load(root / "campaign/production-campaign-manifest.json")
    evidence = load(root / "evidence-summary.json")
    hidden = load(root / "hidden-rule-release-evidence.json")
    qualification = load(root / "qualification/candidate-qualification-evidence.json")
    qualification_run = load(root / "qualification/candidate-qualification-run.json")
    catalog = load(root / "profiles.json")

    verify_artifact_identities(campaign, evidence, hidden, qualification, qualification_run, require)

    expected = {
        "campaignManifestHash": campaign.get("contentHash"),
        "profileCatalogHash": sha256(read_bytes(root / "profiles.json")),
        "evidenceHash": evidence.get("evidenceHash"),
        "hiddenRuleEvidenceHash": hidden.get("evidenceHash"),
        "qualificationEvidenceHash": qualification.get("contentHash"),
        "matrixHash": matrix["contentHash"],
    }
    for field, value in expected.items():
        require(run[field] == value, "run " + field + " artifact binding differs")
    campaign_hash = run["campaignManifestHash"]
    for label, document in (("campaign evidence", evidence), ("qualification evidence", qualification),
                            ("qualification run", qualification_run)):
        require(document.get("campaignManifestHash") == campaign_hash,
                label + " belongs to another campaign")
    require(qualification_run.get("qualificationEvidenceHash") == run["qualificationEvidenceHash"],
            "qualification run evidence binding differs")
    require(matrix["evidenceHash"] == run["evidenceHash"], "matrix campaign evidence binding differs")
    require(matrix["hiddenRuleEvidenceHash"] == run["hiddenRuleEvidenceHash"],
            "matrix hidden-rule evidence binding differs")
    require(run["autonomousCampaignStatus"] == matrix["autonomousCampaignStatus"]
            and run["autonomyClaimAuthorized"] == matrix["autonomyClaimAuthorized"],
            "run autonomy summary differs from matrix")

    verify_matrix_profiles(matrix, catalog, require)
    require(matrix["contentHash"] == matrix_hash(matrix), "matrix canonical content hash differs")
    require(run["contentHash"] == run_hash(run), "run canonical content hash differs")
